### PR TITLE
[Parse] Provide a diagnostic when a closure parameter is declared with type sugar

### PR DESCRIPTION
**Pull request by tapthaker**
Closure parameters are not allowed to be shadow sugared types. This PR tries to provide an error diagnostic for the same.

Resolves #54133.

### CODE REVIEW

**Comment by tapthaker** [comment]
@CodaFi  A few questions regarding the PR:

1. Can you explain why closure parameters are not allowed to be shadow sugared types in the first place?
2. Are the optional tokens ( '?'  &  '!'  ) the only things that should be checked for "sugared types"?

**Comment by CodaFi** [comment]
1. It’s an oversight in the logic we use for parsing parameters that this was ever allowed.  We’ve historically been bad about this - the two-phase needlessly abstract parameter parsing logic is to blame in large part.

2. Checking the type repr kind is the right idea if you want to emit a tailored diagnostic.  You can use `isa<T>` instead of asking for the kind directly.

There’s a third concern here which is that this change is source-breaking. I am aware of at least one client that uses `[Foo]` as a type in parameter position in an overload set.  This patch breaks them.  I think we need to go to Swift Evolution about that last part, which is why I say this isn’t *quite* a starter bug.

File: test/Parse/closure_params.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// RUN: %target-swift-frontend -typecheck -verify %s
+
+struct Bar {}
+
```
Comment (CodaFi) [inline test/Parse/closure_params.swift]: I’ve committed this test [already](https://github.com/apple/swift/blob/5a1cae119ef283508f7a3218d251639a88196822/test/decl/func/functions.swift#L190).  Could you rerun the test suite and update the diagnostics there?

Comment (tapthaker) [reply test/Parse/closure_params.swift]: Done

**Comment by CodaFi** [comment]
Remember that overload resolution is diagnosing this situation today.  The non-overloaded case sometimes compiles today.  The test only contains the overloaded case.

**Comment by tapthaker** [comment]
Got it, I didn't think about that. Cool, so I will start a discussion on swift-evolution soon.

File: lib/Parse/ParsePattern.cpp [hunk 2]
```diff
@@ -373,8 +377,16 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
             // SE-110 depending on the kind of declaration.  We now need to
             // warn about the misuse of this syntax and offer to
             // fix it.
-            diagnose(typeStartLoc, diag::parameter_unnamed_warn)
-              .fixItInsert(typeStartLoc, "_: ");
+            // An exception to this rule is when the type is declared with type sugar
+            // Reference: SR-11724
+            if ( isa<OptionalTypeRepr>(param.Type)
```
Comment (CodaFi) [inline lib/Parse/ParsePattern.cpp]: Nit: A little bit of extra whitespace seems to have slipped in.

**Comment by CodaFi** [comment]
Just one tiny little nit and I think this should be good to take.  Thanks!

**Comment by tapthaker** [comment]
> Remember that overload resolution is diagnosing this situation today. The non-overloaded case sometimes compiles. The test only contains the overloaded case.

@CodaFi:  I am unable to create an example where the non-overloaded case would compile. Could you give it a try?

My example:

```swift
struct Bar {}

struct Foo {
  func registerCallback(_ callback: @escaping (Bar?) -> Void) {}
}

Foo().registerCallback { (Bar?) in }
```
Fails with:

```bash
/tmp/Hello.swift:9:30: error: expected ',' separator
Foo().registerCallback { (Bar?) in }
                             ^
                             ,
/tmp/Hello.swift:9:30: error: expected parameter name followed by ':'
Foo().registerCallback { (Bar?) in }
```

File: lib/Parse/ParsePattern.cpp (hunk 2 above)
Comment (tapthaker) [reply lib/Parse/ParsePattern.cpp]: Done

**Comment by CodaFi** [comment]
Oh, sorry, let me try to expand here:

The source break stems from the general fix to this bug, which is to make what is currently a warning in that test you just updated a hard error.  There are known cases where type sugar in parameter position happens to compile, and the most egregious one I know of uses array sugar: https://swift.godbolt.org/z/gkqhCw.  Even more distressingly, that one also accepts `[Bar?]`

Getting this through source compatibility suite was insufficient to catch users of this the last time I tried, so my general fear is that we'll regress *again*.  We need to go to evolution to get an up-or-down vote on whether we should be allowed to ban this, which will give us a clearer picture of whether we need to go fixup the failing projects, or continue to warn about them until we add a new language version or whatever.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
Hm, these jobs never finished

@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
You got a bad builder on macOS

@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by CodaFi** [comment]
⛵️

**Comment by tapthaker** [comment]
@CodaFi did you folks internally conclude to merge this feature ?

Also, I am sorry I couldn’t spend enough time to create a Swift-evolution discussion around this.

**Comment by CodaFi** [comment]
I merged here because we’ve passed the 5.2 rebranch date. If this breaks any more adopters then we’ll have time to fix them and assess the scope of the issue.  We still need to go to evolution about the broader underlying issue.

**Comment by tapthaker** [comment]
@CodaFi 
Stated a discussion in the swift forums: [suggestion-dont-allow-closure-parameters-to-shadow-types](https://forums.swift.org/t/suggestion-dont-allow-closure-parameters-to-shadow-types/31639). Feel free to add more details there.

**Pull request merged**